token rounding error

ssardana08 opened the pull request:
## Fixes
- trnasaction.Value rename to ValueZCN to be more readable

## Changes
- new package `currency` is introduced
- `currency` will have all implementation related to 0chain tokens (ZCN, mZCN, uZCN and SAS)
- DistributeRewards corrected to ensure 0 coin loss.
- Blobber Block Reward distribution corrected to ensure 0 coin loss.

## Need to be mentioned in CHANGELOG.md?

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

ssardana08 committed:
- Currency coin implemented

ssardana08 committed:
- msgpack updated

ssardana08 committed:
- go mod updated

ssardana08 committed:
- msgpack updated

ssardana08 committed:
- moved equallyDistributeRewards to stakepool

ssardana08 committed:
- DistributeReward fixed

ssardana08 committed:
- blobber block rewards fixed with distribution

ssardana08 committed:
- lint errors handled

ssardana08 committed:
- lint errors handled

ssardana08 committed:
- adjust challenge pool fixed

ssardana08 committed:
- lint errors

ssardana08 committed:
- lint errors

ssardana08 committed:
- lint errors

ssardana08 committed:
Merge branch 'staging' into issue319

ssardana08 committed:
- handler fixed

ssardana08 committed:
- package change currency

ssardana08 committed:
- currency methods ParseZCN used

ssardana08 committed:
- valueZCN set to 1ZCN

ssardana08 committed:
- zcn methods implemented

ssardana08 committed:
- fauscet getconfig updated to currency methods

ssardana08 committed:
- fix faucet handler

ssardana08 committed:
- blockrewards edge case handled
- equaldistribution created a method to sp

ssardana08 committed:
Merge branch 'staging' into issue319

ssardana08 committed:
- ParseZCN implemented

ssardana08 committed:
- minersc config updated

ssardana08 committed:
- writepool min lock data type fixed

ssardana08 committed:
- testcase fixed

ssardana08 committed:
- writepool tests fixed

ssardana08 committed:
- lint errors fixed

peterlimg commented on code/go/0chain.net/chaincore/currency/conversion.go:
```diff
@@ -0,0 +1,152 @@
+	return b, nil
+}
+
+// MultUint64 multiplies Coin a by b, returning an error if the values overflow
```
Adjust the comment, it's actually multiplies Coin c by b.

peterlimg commented on code/go/0chain.net/chaincore/currency/conversion.go:
```diff
@@ -0,0 +1,152 @@
+	return Coin(a), nil
+}
+
+// AddCoin adds a and b, returning an error if the values overflow
```
Adjust comment

peterlimg commented on code/go/0chain.net/chaincore/currency/conversion.go:
```diff
@@ -0,0 +1,152 @@
+package currency
```
Rename the file name to currency.go

peterlimg commented on code/go/0chain.net/chaincore/chain/state/state_context.go:
```diff
@@ -233,7 +235,7 @@ func (sc *StateContext) GetEventDB() *event.EventDb {
+	var amount currency.Coin
 	for _, transfer := range sc.transfers {
 		if transfer.ClientID == sc.txn.ClientID {
 			amount += transfer.Amount
```
Call amount.AddCoin() and check error

peterlimg commented on code/go/0chain.net/chaincore/chain/state/state_context.go:
```diff
@@ -242,13 +244,10 @@ func (sc *StateContext) Validate() error {
+	totalValue := currency.Coin(sc.txn.ValueZCN)
 	if config.DevConfiguration.IsFeeEnabled {
-		totalValue += state.Balance(sc.txn.Fee)
+		totalValue += currency.Coin(sc.txn.Fee)
```
Call totalValue.AddCoin() and check error

peterlimg commented on code/go/0chain.net/chaincore/tokenpool/lock_pool_test.go:
```diff
@@ -277,22 +279,22 @@ func TestZcnLockingPool_FillPool(t *testing.T) {
 
 	p := tokenpool.ZcnPool{}
-	p.Balance += state.Balance(txn.Value)
+	p.Balance += currency.Coin(txn.ValueZCN)
```
Call p.Balance.AddCoin() and check error

peterlimg commented on code/go/0chain.net/chaincore/tokenpool/zcn_pool.go:
```diff
@@ -37,37 +39,37 @@ func (p *ZcnPool) GetID() string {
-	p.Balance += state.Balance(txn.Value)
-	tpr := &TokenPoolTransferResponse{TxnHash: txn.Hash, FromClient: txn.ClientID, ToPool: p.ID, ToClient: txn.ToClientID, Value: state.Balance(txn.Value)}
-	transfer := state.NewTransfer(txn.ClientID, txn.ToClientID, state.Balance(txn.Value))
+	p.Balance += currency.Coin(txn.ValueZCN)
```
p.Balance.AddCoin() and check error

peterlimg commented on code/go/0chain.net/chaincore/transaction/entity.go:
```diff
@@ -61,7 +61,7 @@ type Transaction struct {
 	ChainID         string           `json:"chain_id,omitempty" msgpack:"chid"`
 	TransactionData string           `json:"transaction_data" msgpack:"d"`
-	Value           int64            `json:"transaction_value" msgpack:"v"` // The value associated with this transaction
+	ValueZCN        int64            `json:"transaction_value" msgpack:"v"` // The value associated with this transaction
```
We can keep the name as Value and use SAS as unit. Otherwise, the transaction.Value would not be able to present float tokens such as 1.5 ZCN, which is pretty normal value that could be represented by SAS as 1.5e10. Also, transaction.Fee is SAS. So we'd better change it to have the same unit. Anyway, this will require changes from gosdk too.

ssardana08 replied:
yes. so for the time being we can change it to Value and keep it as int64
I'll create a new ticket to analyse how this is used.. looks like it is used as ZCN at some places and SAS as other.
We'll have to dig deeper in it.

peterlimg commented on code/go/0chain.net/chaincore/wallet/wallet_test.go:
```diff
@@ -188,7 +190,7 @@ func generateTransactions(mpt util.MerklePatriciaTrieI, wallets []*Wallet, trans
 					panic(err)
 				}
-				s.Balance -= state.Balance(value)
+				s.Balance -= currency.Coin(value)
```
Add a Coin.MinusCoin() method that do the coin minus, check if the coin will greater than the origin value to see if there's overflow happens, and return err if any.

peterlimg commented on code/go/0chain.net/chaincore/wallet/wallet_test.go:
```diff
@@ -203,7 +205,7 @@ func generateTransactions(mpt util.MerklePatriciaTrieI, wallets []*Wallet, trans
 				panic(err)
 			}
-			s.Balance += state.Balance(value)
+			s.Balance += currency.Coin(value)
```
Call AddCoin() method  and check error. Perhaps the `AddCoin()` could be renamed to `Add()` to reduce the repeat naming.

peterlimg commented on code/go/0chain.net/smartcontract/minersc/fees.go:
```diff
@@ -397,7 +399,7 @@ func (msc *MinerSmartContract) sumFee(b *block.Block,
 		feeStats.Inc(totalMaxFee)
 	}
-	return state.Balance(totalMaxFee)
+	return currency.Coin(totalMaxFee)
```
The above totalMaxFee += txn.Fee may need negative fee checking, or we should check if txn.Fee is <  0 in transaction validation code.

peterlimg commented on code/go/0chain.net/smartcontract/minersc/fees.go:
```diff
@@ -445,11 +447,15 @@ func (msc *MinerSmartContract) payShardersAndDelegates(
 		partf = float64(fee) / float64(len(sharders))
 		partm = float64(mint) / float64(len(sharders))
```
`fee` and `mint` are currency.Coin, i.e., SAS as unit, they can not be break anymore. So what we should do here is like:

```go
sn := len(sharders)

feeShare := fee / sn 
feeLeft := fee % sn

mintShare := mint / sn
mintLeft := mint % sn

distribute the (feeShare + mintShare) to each sharder if it is > 0, then distribute the coins := feeLeft + mintLeft. Have a loop to distribute 1 to each sharder till all coins are gone. 

You can optimize the code structure to reduce the times of DistributeRewards() function calling by preparing all the coins that need to be distribute to each sharder in the way described above, then loop the sharders to run DistributeRewards()  methods.
```

peterlimg commented on code/go/0chain.net/smartcontract/stakepool/stakepool.go:
```diff
@@ -179,58 +182,133 @@ func (sp *StakePool) MintRewards(
+	}
+	if serviceCharge > 0 {
+		reward := serviceCharge
 		sp.Reward += reward
```
Call AddCoin() to add coin

peterlimg commented on code/go/0chain.net/smartcontract/stakepool/stakepool.go:
```diff
@@ -179,58 +182,133 @@ func (sp *StakePool) MintRewards(
+		} else {
+			valueBalance -= reward
+		}
 		pool.Reward += reward
```
Call AddCoin() to add coins and check error

peterlimg commented on code/go/0chain.net/smartcontract/stakepool/stakepool.go:
```diff
@@ -179,58 +182,133 @@ func (sp *StakePool) MintRewards(
 }
 
 func (sp *StakePool) DistributeRewards(
```
Add unit tests to cover different distribute cases

peterlimg commented on code/go/0chain.net/smartcontract/stakepool/stakepool.go:
```diff
@@ -179,58 +182,133 @@ func (sp *StakePool) MintRewards(
 	providerType spenum.Provider,
 	balances cstate.StateContextI,
-) error {
+) (err error) {
```
Add unit tests to cover the different distribute cases

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/allocation.go:
```diff
@@ -1646,16 +1636,20 @@ func (sc *StorageSmartContract) finishAllocation(
 			}
-			d.Spent += state.Balance(reward)
-			d.FinalReward += state.Balance(reward)
+			d.Spent += reward
```
Call AddCoin(), and check error, all coin plus action should be done by calling the AddCoin()

ssardana08 committed:
- writepool tests fixed

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/allocation.go:
```diff
@@ -1684,7 +1678,7 @@ func (sc *StorageSmartContract) finishAllocation(
 		}
 	}
-	cp.Balance -= state.Balance(passPayments)
+	cp.Balance -= currency.Coin(passPayments)
```
Same as AddCoin, add MinusCoin(), Minus() perhaps, and check overflow.

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -164,6 +178,38 @@ func (ssc *StorageSmartContract) blobberBlockRewards(
+				}
+			}
+
+			rl := rewardBal - rShare
```
rl := rewardBal % len(stakePools)

or rl := rewardBal - rShare * len(stakePools)

ssardana08 committed:
Merge branch 'staging' into issue319

ssardana08 committed:
- currency file rename
- comments fixed

ssardana08 committed:
- AddCoin used to add to coins
- reverted name to txn.Value

ssardana08 committed:
- currency add/minus method implemented
- sumfee error added

ssardana08 committed:
- payShardersAndDelegates refactored

ssardana08 committed:
- stakepool currency code improved

ssardana08 committed:
- currency operations fixed

ssardana08 committed:
Merge branch 'staging' into issue319

ssardana08 committed:
- removed from allocation pool of readpool only the portion removed and not complete value

ssardana08 committed:
- finishAllocation refactored
- in finishallocation move to wp only if cp balance is > 0
- bbr distribution code optimised (more readable)

ssardana08 committed:
- finish alloc int64 error ignore

ssardana08 committed:
- splitByShareRatio return fixed

ssardana08 committed:
- error handling improved in blobberReward/blobberPenalty

ssardana08 committed:
- error handling on adding to stakepool reward

ssardana08 committed:
- stakepool DistributeRewards unit test
